include/sys/emmc.h

```diff
@@ -0,0 +1,279 @@
+#ifndef _SYS_EMMC_H_
+#define _SYS_EMMC_H_
+
+#include <inttypes.h>
+#include <sys/mimiker.h>
+#include <sys/device.h>
+#include <stddef.h>
+
+typedef enum emmc_op_mode {
+  EMMCMODE_BOOT,
+  EMMCMODE_CARD_ID,
+  EMMCMODE_INTR,
+  EMMCMODE_TRANSFER,
+  EMMCMODE_INA,
+} emmc_op_mode_t;
+
+typedef enum emmc_result {
+  EMMC_OK,
+  EMMC_ERR_IERR,
+  EMMC_ERR_TIMEOUT,
+  EMMC_ERR_NOT_RESPONDING,
+  EMMC_ERR_INTR,
+  EMMC_ERR_BUSY,
+} emmc_result_t;
+
+typedef enum emmc_resp_type {
+  EMMCRESP_NONE,
+  EMMCRESP_R1,
+  EMMCRESP_R2,
+  EMMCRESP_R3,
+  EMMCRESP_R4,
+  EMMCRESP_R5,
+} emmc_resp_type_t;
+
+typedef struct emmc_response {
+  emmc_resp_type_t type;
+  union {
+    struct {
+      uint32_t card_status;
+      uint8_t cmd_idx;
+    } r1;
+    struct {
+      uint64_t cid_csd[2];
+    } r2;
+    struct {
+      uint32_t ocr;
+    } r3;
+    struct {
+      uint16_t rca31_16;
+      uint8_t status15;
+      uint8_t reg_addr14_8;
+      uint8_t read7_0;
+    } r4;
+    struct {
+      uint16_t win_rca31_16;
+      uint16_t aux;
+    } r5;
+  } response;
+} emmc_response_t;
+
+typedef enum emmc_command {
+  EMMC_CMD_GO_IDLE = 0,
+  EMMC_CMD_SEND_OP_COND = 1,
+  EMMC_CMD_ALL_SEND_CID = 2,
+  EMMC_CMD_SET_RELATIVE_ADDR = 3,
+  EMMC_CMD_SET_DSR = 4,
+  EMMC_CMD_SLEEP_AWAKE = 5,
+  EMMC_CMD_SWITCH = 6,
+  EMMC_CMD_SELECT_CARD = 7,
+  EMMC_CMD_SEND_EXT_CSD = 8,
+  EMMC_CMD_SEND_CSD = 9,
+  EMMC_CMD_SEND_CID = 10,
+  EMMC_CMD_READ_DAT_UNTIL_STOP = 11,
+  EMMC_CMD_STOP_TRANSMISSION = 12,
+  EMMC_CMD_SEND_STATUS = 13,
+  EMMC_CMD_BUSTEST_R = 14,
+  EMMC_CMD_GO_INACTIVE_STATE = 15,
+  EMMC_CMD_BUSTEST_W = 19,
+  EMMC_CMD_SET_BLOCKLEN = 16,
+  EMMC_CMD_READ_BLOCK = 17,
+  EMMC_CMD_READ_MULTIPLE_BLOCKS = 18,
+  EMMC_CMD_WRITE_DAT_UNTIL_STOP = 20,
+  EMMC_CMD_SET_BLOCK_COUNT = 23,
+  EMMC_CMD_WRITE_BLOCK = 24,
+  EMMC_CMD_WRITE_MULTIPLE_BLOCKS = 25,
+  EMMC_CMD_PROGRAM_CID = 26,
+  EMMC_CMD_PROGRAM_CSD = 27,
+  EMMC_CMD_SET_WRITE_PROT = 28,
+  EMMC_CMD_CLR_WRITE_PROT = 29,
+  EMMC_CMD_SEND_WRITE_PROT = 30,
+  EMMC_CMD_SEND_WRITE_PROT_TYPE = 31,
+  EMMC_CMD_ERASE_GROUP_START = 35,
+  EMMC_CMD_ERASE_GROUP_END = 36,
+  EMMC_CMD_ERASE = 38,
+  EMMC_CMD_FAST_IO = 39,
+  EMMC_CMD_GO_IRQ_STATE = 40,
+  EMMC_CMD_LOCK_UNLOCK = 42,
+  EMMC_CMD_APP_CMD = 55,
+  EMMC_CMD_GEN_CMD = 56,
+} emmc_command_t;
+
+#define EMMC_I_DATA_DONE 0x01
+#define EMMC_I_WRITE_READY 0x02
+#define EMMC_I_READ_READY 0x04
+
+typedef uint64_t emmc_wait_flags_t;
```

Comment:
Why not enum? And in case it can't be, why not `uint8_t`?

Reply:
It was declared as `uint64_t` because I wanted to make it a property that could be accessed using `emmc_get_prop` method. This was a design that resulted from an idea of giving an interrupt resource to the child of e.MMC bus. But this idea is gone now.

I'll change it to an enum. I would've done so already, but looks like I missed it in the last update.

Reply:
Ok, it's fixed now.